services/automerge/automerge.go

```diff
@@ -0,0 +1,249 @@
+	pull_model "code.gitea.io/gitea/models/pull"
+	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/graceful"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/process"
+	"code.gitea.io/gitea/modules/queue"
+	pull_service "code.gitea.io/gitea/services/pull"
+)
+
+// prAutoMergeQueue represents a queue to handle update pull request tests
+var prAutoMergeQueue queue.UniqueQueue
+
+// Init runs the task queue to that handles auto merges
+func Init() error {
+	prAutoMergeQueue = queue.CreateUniqueQueue("pr_auto_merge", handle, "")
+	if prAutoMergeQueue == nil {
+		return fmt.Errorf("Unable to create pr_auto_merge Queue")
+	}
+	go graceful.GetManager().RunWithShutdownFns(prAutoMergeQueue.Run)
+	return nil
+}
+
+// handle passed PR IDs and test the PRs
+func handle(data ...queue.Data) []queue.Data {
+	for _, d := range data {
+		id, _ := strconv.ParseInt(d.(string), 10, 64)
+		handlePull(id)
+	}
+	return nil
+}
+
+func add2Queue(pr *models.PullRequest, sha string) {
+	if err := prAutoMergeQueue.PushFunc(strconv.FormatInt(pr.ID, 10), func() error {
+		log.Trace("Adding pullID: %d to the pull requests patch checking queue with sha %s", pr.ID, sha)
+		return nil
+	}); err != nil {
+		log.Error("Error adding pullID: %d to the pull requests patch checking queue %v", pr.ID, err)
+	}
+}
+
+// ScheduleAutoMerge if schedule is false and no error, pull can be merged directly
+func ScheduleAutoMerge(ctx context.Context, doer *user_model.User, pull *models.PullRequest, style repo_model.MergeStyle, message string) (scheduled bool, err error) {
+	lastCommitStatus, err := pull_service.GetPullRequestCommitStatusState(ctx, pull)
+	if err != nil {
+		return false, err
+	}
+
+	// we don't need to schedule
+	if lastCommitStatus.IsSuccess() {
+		return false, nil
+	}
+
+	return true, pull_model.ScheduleAutoMerge(ctx, doer, pull.ID, style, message)
+}
+
+// MergeScheduledPullRequest merges a previously scheduled pull request when all checks succeeded
+func MergeScheduledPullRequest(ctx context.Context, sha string, repo *repo_model.Repository) error {
+	pulls, err := getPullRequestsByHeadSHA(ctx, sha, repo, func(pr *models.PullRequest) bool {
+		return !pr.HasMerged && pr.CanAutoMerge()
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, pr := range pulls {
+		add2Queue(pr, sha)
+	}
+
+	return nil
+}
+
+func getPullRequestsByHeadSHA(ctx context.Context, sha string, repo *repo_model.Repository, filter func(*models.PullRequest) bool) (map[int64]*models.PullRequest, error) {
+	gitRepo, err := git.OpenRepository(ctx, repo.RepoPath())
+	if err != nil {
+		return nil, err
+	}
+	defer gitRepo.Close()
+
+	refs, err := gitRepo.GetRefsBySha(sha, "")
+	if err != nil {
+		return nil, err
+	}
+
+	pulls := make(map[int64]*models.PullRequest)
+
+	for _, ref := range refs {
+		// If the branch starts with "pull/*" we know we're dealing with a fork.
```

Comment:
this isn't true. Every pull gets a /refs/pull/*/head ref even non-forked repos so every pull will be doubled in this list.

Reply:
@zeripath can we relay on this assumption? bzw. is it written down so it wont get broken on a refactor?

Reply:
I think you can rely on this assumption as it's deep within the code and the meaning of the /refs/pull/*/head.

Reply:
ok I'll remove the other routine ...

Reply:
[8df24af](https://github.com/go-gitea/gitea/pull/9307/commits/8df24afa46c006cba78bc3a70651d2a88a05b375)